### PR TITLE
Use range-based for loops in the generator.

### PR DESCRIPTION
**Pull request by benjaminp**
This makes the code cleaner and also fixes a bunch of signedness warnings from GCC.

Due to our limited support resources, we are focusing our attention on small bug fixes that have limited compatibility concerns. Most other PRs will be rejected.

### CODE REVIEW

File: generator/js_generator.cc [hunk 1]
```diff
@@ -263,14 +263,14 @@ char ToLowerASCII(char c) {
 std::vector<std::string> ParseLowerUnderscore(const std::string& input) {
   std::vector<std::string> words;
   std::string running = "";
-  for (int i = 0; i < input.size(); i++) {
-    if (input[i] == '_') {
+  for (auto c : input) {
```
Comment (dibenede) [inline generator/js_generator.cc:266]: style tweak: use const auto& or auto& instead to explicitly signal as appropriate.

File: generator/js_generator.cc [hunk 2]
```diff
@@ -3446,72 +3442,72 @@ void Generator::GenerateExtension(const GeneratorOptions& options,
 bool GeneratorOptions::ParseFromOptions(
     const std::vector<std::pair<std::string, std::string> >& options,
     std::string* error) {
-  for (int i = 0; i < options.size(); i++) {
-    if (options[i].first == "add_require_for_enums") {
-      if (options[i].second != "") {
+  for (auto option : options) {
```
Comment (dibenede) [inline generator/js_generator.cc:3445]: optional: bonus points for destructuring instead

File: generator/js_generator.cc (hunk 1 above)
Comment (benjaminp) [reply generator/js_generator.cc:266]: Maybe it doesn't matter, but that does not seem to be the existing style of this file. E.g.:
https://github.com/protocolbuffers/protobuf-javascript/blob/3ff6090f139d71453062fb96c66e9aff801709c2/generator/js_generator.cc#L391 https://github.com/protocolbuffers/protobuf-javascript/blob/3ff6090f139d71453062fb96c66e9aff801709c2/generator/js_generator.cc#L397 https://github.com/protocolbuffers/protobuf-javascript/blob/3ff6090f139d71453062fb96c66e9aff801709c2/generator/js_generator.cc#L1762 https://github.com/protocolbuffers/protobuf-javascript/blob/3ff6090f139d71453062fb96c66e9aff801709c2/generator/js_generator.cc#L3355

In general, this we find ourselves iterating over integers or pointers, which I think should be fine to iterate by value?

File: generator/js_generator.cc (hunk 2 above)
Comment (benjaminp) [reply generator/js_generator.cc:3445]: Is it okay to require C++17?

Comment (dibenede) [reply generator/js_generator.cc:3445]: Good question. Disregard the destructing suggestion.  Let's just change this to const auto& option.

File: generator/js_generator.cc (hunk 1 above)
Comment (dibenede) [reply generator/js_generator.cc:266]: Generally speaking, our style is to add & for object types. I think I misread the thing I tagged this on. I meant to do it on something that was a std::string. Similar, we should annotate pointers with auto*|const auto*.

Technically speaking, I think auto should do the right thing most of the time. It's just useful context to the reader and maybe an occasional nudge to auto.

I didn't appreciate that the surrounding code may not follow this guidance, so I'm ok taking the changes as is.

**Review by dibenede — APPROVED** [review]

**Pull request merged**